File: typescript/src/failure-log.ts
/**
 * Rate-limited failure logging. Logs the first occurrence of a failure kind
 * immediately, then suppresses repeats within a quiet window and reports how
 * many were suppressed on the next log line for that kind.
 */
export class FailureLogger {
  private readonly lastLogged = new Map<string, number>();
  private readonly suppressed = new Map<string, number>();

  constructor(
    private readonly quietMs: number = 60_000,
    private readonly clock: () => number = Date.now,
    private readonly sink: (message: string) => void = (m) => console.warn(m),
  ) {}

  report(kind: string, message: string): void {
    const now = this.clock();
    const last = this.lastLogged.get(kind);
    if (last !== undefined && now - last < this.quietMs) {
      this.suppressed.set(kind, (this.suppressed.get(kind) ?? 0) + 1);
      return;
    }
    const suppressedCount = this.suppressed.get(kind) ?? 0;
    this.suppressed.delete(kind);
    const suffix = suppressedCount
      ? ` (${suppressedCount} more suppressed in the last ${Math.round(this.quietMs / 1000)}s)`
      : "";
    this.sink(`metergraph: ${message}${suffix}`);
    this.lastLogged.set(kind, now);
  }
}

File: python/src/metergraph/_failure_log.py
"""Rate-limited failure logging.

Logs the first occurrence of a failure kind immediately, then suppresses
repeats within a quiet window and reports how many were suppressed on the
next log line for that kind. Keeps log volume bounded under sustained
failure without ever going completely silent.
"""

from __future__ import annotations

import logging
import time
from typing import Callable


log = logging.getLogger("metergraph")


class FailureLogger:
    def __init__(
        self,
        quiet_seconds: float = 60.0,
        *,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        self._quiet_seconds = quiet_seconds
        self._clock = clock
        self._last_logged: dict[str, float] = {}
        self._suppressed: dict[str, int] = {}

    def report(self, kind: str, message: str, level: int = logging.WARNING) -> None:
        now = self._clock()
        last = self._last_logged.get(kind)
        if last is not None and now - last < self._quiet_seconds:
            self._suppressed[kind] = self._suppressed.get(kind, 0) + 1
            return
        suppressed = self._suppressed.pop(kind, 0)
        suffix = (
            f" ({suppressed} more suppressed in the last {self._quiet_seconds:.0f}s)"
            if suppressed
            else ""
        )
        log.log(level, "metergraph: %s%s", message, suffix)
        self._last_logged[kind] = now
